### ISSUE TITLE
[BUG] kubectl drain node gets stuck forever

### ISSUE THREAD

**Issue by Martin-Weiss** [longhorn/longhorn#2673]
**Describe the bug**
In case we want to drain a node (RKE2 1.20.7 rke2r2 / longhorn 1.1.100) the drain gets stuck forever in 

```
evicting pod longhorn-system/instance-manager-r-b4be9e85
error when evicting pods/"instance-manager-r-b4be9e85" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```

**To Reproduce**
deploy rke2 (3 master, >=4 worker)
deploy longhorn
deploy rancher-monitoring which creates two PVCs
kubectl drain on one worker that has a replica of the grafana or prometheus PV

**Expected behavior**
The drain should complete

**Log**
```
evicting pod longhorn-system/instance-manager-r-b4be9e85
error when evicting pods/"instance-manager-r-b4be9e85" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```

You can also attach a *Support Bundle* here. You can generate a Support Bundle using the link at the footer of the Longhorn UI.
--> will attach this in a few minutes.

**Environment:**
 - Longhorn version: 1.1.100
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): rancher catalog app
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE2 1.20.7 rke2r2
   - Number of management node in the cluster: 3
   - Number of worker node in the cluster: 14
 - Node config
   - OS type and version: SLES 15 SP2
   - CPU per node:
   - Memory per node:
   - Disk type(e.g. SSD/NVMe):
   - Network bandwidth between the nodes:
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Baremetal
 - Number of Longhorn volumes in the cluster: 4 (2 active)

**Comment by Martin-Weiss** [comment]
[longhorn-support-bundle_ae9587bb-1aa8-4c1f-93a6-e3dfb3cf7160_2021-06-10T14-01-49Z.zip](https://github.com/longhorn/longhorn/files/6631954/longhorn-support-bundle_ae9587bb-1aa8-4c1f-93a6-e3dfb3cf7160_2021-06-10T14-01-49Z.zip)

**Comment by Martin-Weiss** [comment]
Just one additional observation - we have PVCs in the cluster that are not used at the moment. These PVCs result in detached volumes in Longhorn.
In my lab I could get the drain working on one node after I deleted the PVCs that are linked to the detached volumes.
Could it be that the handling of pod disruption budget is not working as expected in case we have "detached longhorn volumes"?

An other cause could be that we have snapshots and backup enabled.. maybe this also impacts the behavior..

**Comment by cclhsu** [comment]
@Martin-Weiss Currently, I was able to identify several scenarios that would cause PDB.  I saw from log that you have create longhorn-test-pvc-rwx. Is it running at time of draining? For monitoring storage, which storage accessModes are you using?

## Known Scenarios

### Scenario 1: Storage class has `numberOfReplicas` of `1`

- If volume storageclass `numberOfReplicas` is `1`, need to increase the number to 2. Otherwise, will encounter PDB errors during draining and upgrade will time out.

### Scenario 2: PVC/PV/LHV is created through Longhorn UI, but has not yet attached and replicated

- After volume is attached, replicate, and detached, the nodes with volume replicas can be drained successfully.
- This issue does not seem to be an issue if volume create through pvc using manifest.

### Scenario 3: PVC/PV/LHV is created through Longhorn UI and attached to a host node

- Need to detach then volume can be drain successfully.

### Scenario 4: RWX volume attached to a node

- Need to scale down the workload and drain.

### Scenario 5: RWO volume with last healthy replica

- set `allow-node-drain-with-last-healthy-replica` to `true` and able to drain.

**Comment by joshimoo** [comment]
For scenario 5 if you do this, you will potentially lose your data, so handle with care.

**Comment by Martin-Weiss** [comment]
> @Martin-Weiss Currently, I was able to identify several scenarios that would cause PDB. I saw from log that you have create longhorn-test-pvc-rwx. Is it running at time of draining? For monitoring storage, which storage accessModes are you using?

Thanks a lot for looking into this issue!

In this cluster I have created two test-pvc - one with RW and one with RWX but both are "detached" and no pod is using these at the time of drain. Other PVCs are RW only.

So why does a "detached volume" block the drain?

Regarding the scenarios listed
1 - numberOfReplicas = 3 - so should not apply in our case
2 - have created the volumes via PVCs in yaml but yes - they might have never been attached somewhere. How can we check if any detached volume is in this state / monitor it / can this be fixed so that "detached" per default does not block a drain? And as we have created the PV via PVC yaml - this should not apply in our case
3 - does not apply in our case
4 - does not apply in our case
5 - does not apply in our case

So based on the listed scenarios only  "2" seems to be something that might be related.. so what is the problem in case such a volume has been created via Longhorn UI and why does the problem not occur in case of yaml based creation? Is this a known issue?

Thanks!!

**Comment by cocampbe** [comment]
I've had this same issue. It's usually due to the instance manager being the last one with valid storage. I even scaled down the deployment that was using storage on that IM. 

When doing node maintenance, ie, os patching, waiting on node drains and replica rebuilds takes forever. I can deal with the drains, but the rebuilds are slow. Outside of finding faster ssd storage, I am not sure how this could be sped

**Comment by PhanLe1010** [comment]
Thanks @Martin-Weiss for reporting

I think this a Longhorn bug:

When the volume is created the first time (the volume could be created via UI or via PVC yaml manifest) and has never been attached to a node, Longhorn doesn't remove the PDB for `the instance-manager-r-xxx` pod that contains the volume's replicas when user run `kubectl drains`. This blocks `kubectl drain` command

The reason Longhorn doesn't remove the PDB is because Longhorn is trying to find a healthy replica on a different node by checking [r.Spec.HealthyAt != ""](https://github.com/longhorn/longhorn-manager/blob/2e8edc6cfd2ed9e8692377e544df8977306428c6/controller/instance_manager_controller.go#L589). This check always fails on the volume that has never been attached to a node since the  `r.Spec.HealthyAt` has never been set.

**Workaround:**

Find the volumes that have never been attached to a node and try attach them then detach them. This will set the `r.Spec.HealthyAt` for the volumes' replicas so Longhorn will remove the PDB for `the instance-manager-r-xxx` pod that contains the volume's replicas when user run `kubectl drains`. To find those volumes, you can run `kubectl get replicas -n longhorn-system -o yaml` and find the replicas that have `failedAt == "" and healthyAt == ""`, and get the volume name from the `replica.metadata.metadata[0].name`

**Comment by PhanLe1010** [comment]
@cocampbe We are working on speeding up the replica rebuilding process. The effort is tracked at https://github.com/longhorn/longhorn/issues/2507
cc @joshimoo

**Comment by joshimoo** [comment]
@cocampbe can you run the below fiemap against one of the replicas snapshot files that take a long time to rebuild.
https://github.com/joshimoo/fiemap/archive/refs/tags/v0.2.0.tar.gz

Build & run Instructions below:
https://github.com/longhorn/longhorn/issues/2507#issuecomment-857195496

That way we can get some details from user clusters in the wild.
Can you also share the kernel version `uname -u` from your cluster nodes.

**Comment by cocampbe** [comment]
kernel is 4.14.35-2025.404.1.1.el7uek.x86_64. Working on getting the other info.

**Comment by cocampbe** [comment]
Hopefully this is what you are looking for.

[volume-snap-c411a9f2-3ace-4cdf-930d-edf76994d81a_fiemap.txt](https://github.com/longhorn/longhorn/files/6664930/volume-snap-c411a9f2-3ace-4cdf-930d-edf76994d81a_fiemap.txt)

**Comment by innobead** [comment]
@cocampbe I would suggest you can create another issue we can tackle your problem specifically because the original issue reported by @Martin-Weiss already identified by https://github.com/longhorn/longhorn/issues/2673#issuecomment-861925034. 

This can help you figure out your case. Thanks.

**Comment by Martin-Weiss** [comment]
> 
> 
> Thanks @Martin-Weiss for reporting
> 
> I think this a Longhorn bug:
> 
> When the volume is created the first time (the volume could be created via UI or via PVC yaml manifest) and has never been attached to a node, Longhorn doesn't remove the PDB for `the instance-manager-r-xxx` pod that contains the volume's replicas when user run `kubectl drains`. This blocks `kubectl drain` command
> 
> The reason Longhorn doesn't remove the PDB is because Longhorn is trying to find a healthy replica on a different node by checking [r.Spec.HealthyAt != ""](https://github.com/longhorn/longhorn-manager/blob/2e8edc6cfd2ed9e8692377e544df8977306428c6/controller/instance_manager_controller.go#L589). This check always fails on the volume that has never been attached to a node since the `r.Spec.HealthyAt` has never been set.
> 
> From your provided support bundle I can see that you are having 2 volumes that have never been attached.
> 
> **Workaround:**
> 
> Find the volumes that have never been attached to a node and try attach them then detach them. This will set the `r.Spec.HealthyAt` for the volumes' replicas so Longhorn will remove the PDB for `the instance-manager-r-xxx` pod that contains the volume's replicas when user run `kubectl drains`. To find those volumes, you can run `kubectl get replicas -n longhorn-system -o yaml` and find the replicas that have `failedAt == "" and healthyAt == ""`, and get the volume name from the `replica.metadata.ownerReferences[0].name`

Thanks a lot for the analysis! 

Unfortunately this can happen at any time and we can not include such a test / attach / detach in the automated update&upgrade process for the OS and for RKE2 so this somehow needs to be handled automatically by Longhorn...

Do you have any idea when we can expect this to be fixed / how this can be fixed?

**Comment by innobead** [comment]
@khushboo-rancher we will add this to 1.1.2 as well.

cc @kaxing @meldafrawi

**Comment by khushboo-rancher** [comment]
Verified with Longhorn-v1.1.2-head and Longhorn-master `07/28/2021`

Validation - **Pass**

I was able to reproduce the problem of `Scenario 2: PVC/PV/LHV is created through Longhorn UI, but has not yet attached and replicated` with Longhorn-v1.1.1

With Longhorn-v1.1.2-head and master, the node is able to get drained successfully if there is a volume which never got attached to a node.

**Comment by wirwolf** [comment]
Hello @cclhsu. A want to discuss Scenario 1. Why does Longhorn not move volume to another node(in my test he tries, but he delete new volume after rebuild)?

**Comment by mantissahz** [comment]
Failed to drain the node with v1.3.0-rc3
Stuck at
`“error when evicting pods/"instance-manager-r-dac5b6e6" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget`

**Comment by innobead** [comment]
@mantissahz please don't remove the previous flag. The issue is a regression in 1.3.0-rc instead only. Good catch!

**Comment by innobead** [comment]
Closing this. @mantissahz Please help create another issue to track. Thank you!
